MetaMask/metamask-extension
# feat(offscreen): adds offscreen document

## **Description**
For MV3 snaps and Hardware wallet functionality we need to run some sets of code in an offscreen document that can communicate with the extension. The purpose of this Pull request is to start the conversation about how to build this. This PR should be a stepping stone for the fundamentals of #17687 and #17008 to incrementally be adopted. 

The idea here is that we create a new "entry point" into the application in the way of this offscreen.html file. There is a offscreen.ts file that can be used to initialize a PostMessageStream for the iframe itself, which is done in #17008. Essentially the snaps-execution-environment offscreen document sets up lockdown and then initializes the OffscreenSnapExecutor with its PostMessageStream. These pieces can be added in a future PR, although I might need some guidance on the proper way to setup lockdown.

In terms of Hardware wallet support, the iframes in #17687 and the supporting typescript files can be included in layers ontop of this folder structure, eventually getting to full hardware wallet support but we should tackle these one at a time. There are a couple PRs opened by @mikesposito to upgrade trezor and ledger keyrings that are also piecing out elements from #17687 that are prereqs for continuing HW support in MV3. 

This uses post-message-stream which will later be used as well for snaps execution in the offscreen document. 

## **Related issues**
Partially progresses #21622 

## **Manual testing steps**

1. run `yarn start:mv3`
2. Go to extensions in chrome.
3. You should see both a service-worker and offscreen.html file loaded and be able to inspect both.
4. Open both windows. From the service-worker console, run `chrome.runtime.sendMessage({target: 'child', data: 'test'})`
5. See that the offscreen.html console logs the message. 
6. From the offscreen console, run `chrome.runtime.sendMessage({target: 'parent', data: 'test'})`
7. See that the service worker console logs the message. 

## **Screenshots/Recordings**



### **Before**



### **After**



## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
